ENH: sparse: add scipy.sparse.linalg.spsolve_triangular

jor- opened the pull request:
I would like to add the `scipy.sparse.linalg.spsolve_triangular` function. It solves linear equations `A x = b` for `x` with sparse triangular matrix `A`. Its API is inspired by `scipy.linalg.solve_triangular`. Tests are also included.

pv commented:
Some general comments:
- Is adding this important? Does it eg outperform superlu? Does superlu have triangular solve?
- The for loops probably can and should be vectorized, and one can ask e.g. if implementing it in sparsetools would make sense.

jor- commented:
I have vectorized the ``for`` loops.

``SuperLU`` does not have a special function to solve systems with triangle matrices and my proposed function is faster than the general solve ``function`` from ``SuperLU``. See the following measurements.

```
In [57]: n = 10**3
    ...: A = random_triangle_matrix(n)
    ...: m = 10
    ...: b = np.random.rand(n, m)

In [58]: % time x = scipy.sparse.linalg.spsolve(A, b)
CPU times: user 352 ms, sys: 12 ms, total: 364 ms
Wall time: 198 ms

In [59]: % time x = scipy.sparse.linalg.spsolve_triangular(A, b)
CPU times: user 52 ms, sys: 8 ms, total: 60 ms
Wall time: 55.5 ms

In [60]: n = n*2
    ...: A = random_triangle_matrix(n)
    ...: b = np.random.rand(n, m)

In [61]: % time x = scipy.sparse.linalg.spsolve(A, b)
CPU times: user 1.78 s, sys: 16 ms, total: 1.8 s
Wall time: 949 ms

In [62]: % time x = scipy.sparse.linalg.spsolve_triangular(A, b)
CPU times: user 140 ms, sys: 4 ms, total: 144 ms
Wall time: 140 ms
```

In general I would consider a function for solving systems with triangle matrices important because this a common application like for example when you have a LU decomposition, a QR decompostion or a Cholesky decomposition of a matrix and want to solve a corresponding linear system.

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+
+            ## check regularity and triangularity
+            assert np.all(A.indices[indptr_start:indptr_stop-1] <
+                A.indices[indptr_start+1:indptr_stop])     # indices are assumed to be sorted
```
Is this check necessary? `A.sort_indices()` was called above.

jor- replied:
removed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+                    'entry of A is zero!'.format(i))
+            if A.indices[indptr_stop-1] > i:
+                raise LinAlgError('A is no lower triangle matrix, since the '
+                    'entry at ({},{}) is not zero!'.format(i, A.indices[indptr_stop-1]))
```
These error messages need some revision:
 * "A is singular: {}th diagonal is zero"
 * "A is not lower triangular: entry [{},{}] is nonzero"

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+
+    ## init x
+    if not overwrite_b:
+        x = np.empty(b.shape)
```
Why not `x = b.copy()` and avoid the `x[i] = b[i]` below?

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+                A_values_in_row_i = A.data[indptr_start:indptr_stop-1]
+                if x.ndim > 1:
+                    A_values_in_row_i = A_values_in_row_i[:, np.newaxis]
+                x[i] -= np.sum(x[A_column_indices_in_row_i] * A_values_in_row_i, axis=0)
```
Could this be converted to a dot product?

jor- replied:
You are right. I converted it to a ``np.dot`` call:
```
x[i] -= np.dot(x[A_column_indices_in_row_i].T, A_values_in_row_i)
```

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+            indptr_start = indptr_stop
+    else:
+
+        ## fill backward
```
I don't like how this branch is essentially the same as the lower triangular case. Is it possible to move the lower/upper logic out and reuse the core `x[i]` computation for both cases?

jor- replied:
I had the same idea but found no (good) solution for it.

jor- replied:
I merged the two cases now. The result looks good to me. The only drawback is that there is an if else check regarding ``lower`` at each iteration.

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+        Right-hand side matrix in `a x = b`
+    lower : bool, optional
+        Whether `A` is a lower or upper triangle matrix.
+        Default is lower triangle matrix.
```
triangle -> triangular

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+
+def spsolve_triangular(A, b, lower=True, overwrite_b=False):
+    """
+    Solve the equation `A x = b` for `x`, assuming a is a triangular matrix.
```
"assuming A is a"

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+    Raises
+    ------
+    LinAlgError
+        If `A` is singular.
```
or not triangular?

jor- replied:
yes, changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/tests/test_linsolve.py:
```diff
@@ -498,5 +498,52 @@ def worker():
+                A = scipy.sparse.triu(A)
+            A = A.tocsr()
+            for i in range(n):
+                A[i,i] = np.random.rand() + 1
```
To avoid extra conversions between sparse formats:
```python
A = scipy.sparse.random(n, n, density=0.1, format='coo')
if lower:
    A = scipy.sparse.tril(A)
else:
    A = scipy.sparse.triu(A)
A.setdiag(np.random.random(n) + 1)
return A.tocsr(copy=False)
```

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/tests/test_linsolve.py:
```diff
@@ -498,5 +498,52 @@ def worker():
+                A[i,i] = np.random.rand() + 1
+            return A
+
+        for n in (10, 10**2, 10**3):
```
Is it necessary to test with larger matrix sizes here?

jor- replied:
Not necessarily, the matrix should only be large enough, so that it contains several non-zero off-diagonal elements. I could fix n for example to 100 if you would prefer this.

perimosocordiae replied:
Hmm, maybe it's better to generate a static fixture matrix for this, then.

jor- replied:
I am not quite sure what you meant by  "static fixture matrix". Do you mean to fix the random state by ``numpy.random.seed``? Or maybe a (static) class variable?

perimosocordiae replied:
I just meant specifying `A` and `b` explicitly, not relying on random generation at all.

jor- replied:
Ah, okay. I would always prefer randomized testing. This may cover test cases which the programmer does not have in mind explicitly. But maybe the random state should be set explicitly by ``numpy.random.seed`` to make it reproducible.

perimosocordiae replied:
@pv, do you have an opinion on this? There are certainly pros and cons, and I'm not sure what the overall scipy policy is.

e-q replied:
FWIW in `scipy.signal`, we have various tests that generate signals with `np.random.randn` and a fixed seed.

pv replied:
Randomized testing is fine, as long as you set the seed. It may also be often preferable, since it can cover cases you didn't think about, and it's easier to verify the correctness of the tests themselves.

jor- replied:
I left the matrices randomized and set the seed fix.

perimosocordiae commented on scipy/sparse/linalg/dsolve/tests/test_linsolve.py:
```diff
@@ -498,5 +498,52 @@ def worker():
+                for lower in (True, False):
+                    A = random_triangle_matrix(n, lower=lower)
+                    x = spsolve_triangular(A, b, lower=lower)
+                    assert_array_almost_equal(np.dot(A.toarray(), x), b)
```
`assert_array_almost_equal(A.dot(x), b)`

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/tests/test_linsolve.py:
```diff
@@ -498,5 +498,52 @@ def worker():
+        b = array([[2., 0.], [2., 2.]])
+        for matrix_type in (array, csc_matrix, csr_matrix):
+            x = spsolve_triangular(matrix_type(A), b, lower=True)
+            assert_array_almost_equal(np.dot(A, x), b)
```
Using `A.dot(x)` here is better, because it calls the sparse matrix `dot` directly instead of routing through numpy.

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+                if x.ndim > 1:
+                    A_values_in_row_i = A_values_in_row_i[:, np.newaxis]
+                x[i] -= np.sum(x[A_column_indices_in_row_i] * A_values_in_row_i, axis=0)
+            x[i] /= A.data[indptr_stop-1]           # devide by i-th diagonal element of A
```
devide -> divide

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+            'the size of the first dimension of b but the shape of A is '
+            '{} and the shape of b is {}.'.format(A.shape, b.shape))
+
+    ## init x
```
Comments should start with a single `#`, and ideally should be full sentences.

jor- replied:
changed

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+    Notes
+    -----
+    .. versionadded:: 0.20.0
+    """
```
This docstring is causing the TravisCI build to fail. Each of the underlined section headings should have a blank line before them.

jor- replied:
changed

jor- commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+    if A.ndim != 2:
+        raise ValueError('A must have 2 dimensions but its shape is {}.'.format(A.shape))
+    if A.shape[0] != A.shape[1]:
+        raise ValueError('A must be a square matrix but its shape is {}.'.format(A.shape))
```
changed, removed

jor- commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,125 @@ def solve(b):
+        A = csr_matrix(A)
+
+    A.eliminate_zeros()
+    A.sort_indices()
```
I added a ``copy_A=True`` keyword.

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,108 @@ def solve(b):
+    lower : bool, optional
+        Whether `A` is a lower or upper triangular matrix.
+        Default is lower triangular matrix.
+    copy_A : bool, optional
```
It seems a little strange to have `copy_A` and `overwrite_b`. Looking at the [scipy.linalg docs](https://docs.scipy.org/doc/scipy-0.18.1/reference/linalg.html), it seems pretty standard to have `overwrite_a=False, overwrite_b=False`. I know that we're not _really_ overwriting `A`, though, so I'm conflicted.

jor- replied:
Yes, I agree. I would prefer ``overwrite_A=False`` to have it consistent.

jor- replied:
I changed it to ``overwrite_A=False``.

perimosocordiae commented on scipy/sparse/linalg/dsolve/linsolve.py:
```diff
@@ -383,3 +384,108 @@ def solve(b):
+    else:
+        x = b.copy()
+
+    # Chose forward or backward order.
```
Chose -> Choose

jor- replied:
Changed.

pv commented:
Randomized testing is fine, as long as you set the seed. It may also be 
often preferable, since it can cover cases you didn't think about, and 
it's easier to verify the correctness of the tests themselves.

jor- commented:
Any further comments?

@perimosocordiae: I think I addressed all your comments or did I miss any?

One test failed (``test_continuous_basic.test_cont_basic`` in ``scipy.stats``) but this is not related to my commits.

Should I squash the commits into a single one?

Is this PR going into version 0.19 or the following?

jor- commented:
I changed ``versionadded`` to 0.19.

@perimosocordiae: Thank you for your useful comments.

jor- commented:
Is there anything left to do from my side before the PR can be merged?

ev-br commented:
CI failure is spurious. This has been okayed by CJ, no more comments for a week, merging. Thanks @jor-, all.
